src/main/java/nl/b3p/tailormap/api/controller/MapController.java

```diff
@@ -149,48 +158,165 @@ private void getApplicationParams(@NotNull Application a, @NotNull MapResponse m
-                    mapResponse.addBaseLayersItem(appLayer);
+            } else if (parent != null) {
+                treeNodeMap.get(parent.getId()).addChildrenIdsItem(node.getId());
+            }
+        }
+
+        for (StartLayer l : startLayers) {
+            AppLayer appLayer =
+                    new AppLayer()
+                            .id(l.getApplicationLayer().getId())
+                            .layerName(l.getApplicationLayer().getLayerName())
+                            .title(l.getApplicationLayer().getLayerName())
+                            .serviceId(l.getApplicationLayer().getService().getId())
+                            .visible(l.isChecked());
+
+            mapResponse.addAppLayersItem(appLayer);
+
+            // Find where this startLayer is used.
+            // TODO: can a StartLayer be in >1 place?
+            boolean isBackground = false;
+            boolean isForeground = false;
+            String id = String.format("lyr_%d", l.getApplicationLayer().getId());
+            for (Level level : levelSet) {
+                if (level.getLayers().contains(l.getApplicationLayer())) {
+                    if (backgroundSet.contains(level.getId())) {
+                        isBackground = true;
+                    } else {
+                        isForeground = true;
+                    }
+
+                    LayerTreeNode node = treeNodeMap.get(level.getId());
+                    node.addChildrenIdsItem(id);
                 }
-                mapResponse.addServicesItem(s);
+
+                if (isBackground && isForeground) break;
+            }
+
+            if (!isBackground && !isForeground) continue;
+
+            LayerTreeNode node =
+                    new LayerTreeNode()
+                            .id(id)
+                            .name(l.getApplicationLayer().getLayerName())
+                            .appLayerId(l.getApplicationLayer().getId().intValue());
+
+            if (isBackground) mapResponse.addBaseLayerTreeNodesItem(node);
+            if (isForeground) mapResponse.addLayerTreeNodesItem(node);
+        }
+
+        for (StartLayer l : startLayers) {
+
+            GeoService geoService = l.getApplicationLayer().getService();
+            Service s =
+                    new Service()
+                            .url(geoService.getUrl())
+                            .id(geoService.getId())
+                            .name(geoService.getName())
+                            .protocol(Service.ProtocolEnum.fromValue(geoService.getProtocol()))
+                            .capabilities(geoService.getCapabilitiesDoc());
+            if (geoService.getProtocol().equalsIgnoreCase(TileService.PROTOCOL)) {
+                s.tilingProtocol(
+                        Service.TilingProtocolEnum.fromValue(
+                                ((TileService) geoService).getTilingProtocol()));
             }
+            mapResponse.addServicesItem(s);
```

Comment:
there is a small problem in (this already) existing code; it adds the service of each layer to the response, duplicating services, I will push a change to this branch that fixes this.

eg `GET http://snapshot.tailormap.nl/api/app/1/map` will show this

Reply:
https://github.com/B3Partners/tailormap-api/pull/86/commits/6c6653a1c29e58baed028f8398aba687916d6487

Reply:
also added a postgreqsl integration test for this in a5738160d474872266ec396622fe95d0fefbab5d